### ISSUE TITLE
GPU hist fixed-point quantisation can turn tiny positive Hessians into zero curvature

### ISSUE THREAD

**Issue by RAMitchell** [dmlc/xgboost#12265]
# GPU hist fixed-point quantisation can turn tiny positive Hessians into zero curvature

## Summary

In the GPU histogram algorithm, gradient pairs are quantised into fixed-point integers for deterministic accumulation. For Hessians, this can turn a tiny but positive floating-point Hessian into an integer zero when the value is smaller than one fixed-point quantum.

That changes the meaning of the training signal. A positive Hessian, however small, represents defined local curvature. A zero Hessian represents no curvature. The issue here is that the objective did not produce a true zero Hessian; the GPU quantiser did.

For objectives such as logistic and multiclass, XGBoost explicitly clamps Hessians to a small positive value, e.g. `1e-16`, to avoid semantic zero curvature. GPU fixed-point quantisation can undo that by truncating those positive values to integer zero.

## Why this matters

The GPU code uses quantised Hessian zero as a semantic sentinel in multiple places. For example:

- `src/tree/gpu_hist/expand_entry.cuh`: rejects a split if either child has `GetQuantisedHess() == 0`.
- `src/tree/gpu_hist/sampler.cu`: uses `GetQuantisedHess() == 0` to identify rows that were not sampled.
- `src/tree/updater_gpu_hist.cuh`: encodes sampled/not-sampled state from `GetQuantisedHess() != 0`.
- `src/tree/gpu_hist/multi_evaluate_splits.cu`: checks child/sibling Hessian sums for quantised zero.

So this is not only a small numerical perturbation. A positive-curvature training example can become indistinguishable from a true zero-curvature example, an empty child, or an unsampled row.

## Mechanism

The GPU quantiser builds one scale per target. In simplified form:

$$
q_h = \mathrm{int64}(h / u_t)
$$

where $u_t$ is the per-target fixed-point quantum. Any Hessian satisfying:

$$
0 < h < u_t
$$

is quantised to:

$$
q_h = 0
$$

even though the objective produced a positive Hessian.

The quantum is determined by the target Hessian mass and row count:

$$
u_t = \mathrm{CreateRoundingFactor}(\mathrm{sum\\_positive\\_hess}_t, n_\mathrm{rows}) / 2^{62}
$$

On sufficiently large or heterogeneous datasets, $u_t$ can be larger than the objective's positive Hessian floor.

## Evidence

I reproduced the GPU quantiser calculation in Python and measured positive floating-point Hessians that become fixed-point zero.

For CoverType with `multi:softprob`, seed 2, `tree_method=hist`, `device=cuda`, `eta=0.2`, `max_depth=8`, `max_bin=256`:

| round | zeroed positive Hessians | zeroed full `(grad,hess)` pairs | max fixed-point quantum | min float Hessian |
|---:|---:|---:|---:|---:|
| 5 | 1,244 / 3,253,663 | 1,015 / 3,253,663 | 5.684e-14 | 1.000e-16 |
| 10 | 998 / 3,253,663 | 867 / 3,253,663 | 5.684e-14 | 1.000e-16 |
| 20 | 870 / 3,253,663 | 803 / 3,253,663 | 5.684e-14 | 1.000e-16 |
| 50 | 484 / 3,253,663 | 470 / 3,253,663 | 2.842e-14 | 1.000e-16 |
| 100 | 152 / 3,253,663 | 143 / 3,253,663 | 2.842e-14 | 1.000e-16 |

A synthetic mixed easy/hard stress case also reproduces the issue. With `binary:logistic`, `eta=1.0`, half the rows are easy and half intentionally noisy:

| round | zeroed positive Hessians | zeroed full `(grad,hess)` pairs | fixed-point quantum | min float Hessian |
|---:|---:|---:|---:|---:|
| 50 | 108 / 160,000 | 108 / 160,000 | 3.553e-15 | 1.000e-16 |
| 100 | 5,750 / 160,000 | 5,750 / 160,000 | 3.553e-15 | 1.000e-16 |

Small datasets did not reproduce the issue, because their fixed-point quantum was smaller than the objective Hessian floor. So the problem is not universal; it appears when the quantisation scale is coarse enough relative to the tiny positive Hessian tail.

## Proposed fix

When quantising Hessians, preserve positive curvature by mapping any positive floating-point Hessian to at least one fixed-point quantum:

```cpp
auto q_hess = static_cast<int64_t>(hess * to_fixed_point);
if (hess > 0.0 && q_hess == 0) {
  q_hess = 1;
}
```

This should apply to Hessians only. Gradients are signed, and inventing signed gradient mass would be much harder to justify. Hessians are nonnegative curvature terms, and for these objectives the original floating-point value already says the curvature is positive.
